### ISSUE TITLE
non-POJO warning thrown though adaptor converts data to plain object.

### ISSUE THREAD

**Issue by Worthaboutapig** [ractivejs/ractive#2896]
## Description:
Further to #2892 and similar, I'm attempting to achieve the same (I'd like some form of typing, and also fat models); I'm also using a pub/sub scenario to manage my models, which conflicts to an extent with ractive's 'set' (using 'updateModel' seems to be ok so far).
I've implemented an adaptor to map my typed-models to POJOs for the ractive viewmodel data, however, despite this, it (the debug version) is still printing a warning that the data is not a POJO.
Presumably the 'validate' function at 3413 (v0.7.3) is being called before the adaptor is applied.
Should the check not occur after the adaptation?

## Versions affected:
v0.7.3 and 0.8.11 (and probably all others).

## Platforms affected:
Chrome on Win/Linux.

## Reproduction:
https://jsfiddle.net/worthaboutapig/tdm4nruy/

## Code:

**Comment by evs-chris** [comment]
As long as you aren't trying to use it as the _root_ model e.g. `new Ractive({ data: nonPOJO })`, you should be fine using non-POJOs. If your models don't act like POJOs, for instance by having a composite setter that doesn't end up setting the same values as it gets passed, you will probably need an adaptor to handle additional updates. As long as you make your non-POJO models children of a POJO, you should be fine e.g. `new Ractive({ data: { base: nonPOJO } })`.

**Comment by ghost** [comment]
It would be nice to have a minimal reproduction though. Just to be sure this isn't actually a bug in the regular use case. Also, with a reproduction, we could help out in implementation.

**Comment by Worthaboutapig** [comment]
Thought you might say that, I've created a jsFiddle that achieves the solution I think I want with an adaptor; I believe I'm using the adaptor correctly.
The first ractive creates a viewmodel with a non-POJO root data object, the second is identical (assigning the non-POJO as a root data object), but then adapts it to a POJO.
Is an adaptor supposed to do what I expect? i.e. convert the data to a POJO before ractive may do any shenanigans on it... if not... well, I still can't have constructed objects, which would be ~~a PITA~~... very disappointing.

**Comment by Worthaboutapig** [comment]
Having a look at the source (in 0.8.11), I imagine this won't be an easy fix, assuming it's accepted to change.  

The code that prints the warning, runs the check `validate` from `combine$1`, which, in turn, is part of `init` and `construct` where we have:
>// Create a viewmodel
>var viewmodel = new RootModel({
>  adapt: getAdaptors( ractive, ractive.adapt, options ),
>  data: dataConfigurator.init( ractive.constructor, ractive, options ),
>  ractive: ractive
>});

 and so the validation check is run against the raw data, before any adaptations occur.

Am I misunderstanding the benefits of `adaptors`?  Is the data potentially manipulated and damaged (when non-POJO) before the adaptors are run, as that's not what I'd expect.

**Comment by evs-chris** [comment]
Adaptors generally don't apply to the root model because it is less a model and more a data dictionary that can be built up using a component inheritance hierarchy. I don't think we could safely allow for both inherited data and non-POJO classes as the root model. It _is_ possible though to have a non-POJO value adapted one level below root.

If you're not going to use any inheritance or defaulted data, we may be able to add a flag to avoid the abuse of the data object that is passed in.

**Comment by Worthaboutapig** [comment]
Yeah, so it seems if your root is not a POJO, you have to wrap it to make it appear to be one, like:

    function Fish(name) {
      this.name = name;
    };

    var cod = new Fish("cod");

    var ractiveOptions = {
      ...,
      data: {
        model: cod
        }
    };

I suppose it would be too much magic to auto-wrap POJOs in this sort of way, so that you don't have to care?  It means I have to have this additional property-accessor boiler-plate all over the place, which serves no useful purpose (to my code).

**Comment by wearhere** [comment]
@evs-chris I'm confused by

> As long as you aren't trying to use it as the root model e.g. new Ractive({ data: nonPOJO }), you should be fine using non-POJOs

because https://github.com/ractivejs/ractive-adaptors-backbone is the canonical adaptor AFAICT, and that adaptor handles `new Ractive({ data: someBackboneModel })` just fine.

It's really nice to have the Backbone model be the root model because then its attributes are "directly" available to Ractive, no unnecessary "namespace". That is, [the documentation](https://github.com/ractivejs/ractive-adaptors-backbone#usage) suggests 

```js
var user = new Backbone.Model({
  name: 'world'
});

var ractive = new Ractive({
  el: 'main',
  template: '<h1>Hello {{user.name}}!</h1>',
  data: {
    user: user
  }
});
```

but why not just

```js
var ractive = new Ractive({
  el: 'main',
  template: '<h1>Hello {{name}}!</h1>',
  data: user
});
```

?

**Comment by ghost** [comment]
Digging into the past, [Ractive once only supported POJOs and functions as data](https://github.com/ractivejs/ractive/commit/00a527ce9c5f6bb140a498049525eaa6e6378896).

```
-	if ( childValue && typeof childValue !== 'object' && typeof childValue !== 'function') {
 -		throw new TypeError( 'data option must be an object or a function, "' + childValue + '" is not valid' );
 -	}
 +	validate( childValue );
```

Here's [the relevant PR](https://github.com/ractivejs/ractive/pull/1767) that added that warning, [an analysis about non-POJOs](https://github.com/ractivejs/ractive/pull/1767#issuecomment-77077061) and [the resolution](https://github.com/ractivejs/ractive/pull/1767#issuecomment-77186053). In a gist, the PR addresses how instance data is determined when instantiating from subclasses of Ractive.

From what I see, the danger is how the data is generated. See `combine` and `fromProperties` from [this file](https://github.com/ractivejs/ractive/blob/dev/src/Ractive/config/custom/data.js). `combine` is in charge of building the instance's `data` while `fromProperties` is the one that resolves which one should be on it. 

Now if you look at `fromProperties` and how it copies data:

```js
function fromProperties ( primary, secondary ) {
	if ( primary && secondary ) {
		for ( const key in secondary ) {
			if ( !( key in primary ) ) {
				primary[ key ] = secondary[ key ];
			}
		}

		return primary;
	}

	return primary || secondary;
}
```

This would work for POJOs, but this would not work if you had, say, properties on the prototype of your non-POJO or if you had explicit functions as accessors. [Here's an example](https://jsfiddle.net/9xsh53p8/) of an adapted object used as root data of a direct Ractive instance and as root data for a component definition. It works on the direct instance, but it doesn't work on the component instance. When `fromProperties` is called to resolve the data between the instance and it's component definition, it can't copy anything from from the component definition since `Box` is all accessors. Only when the adapted data becomes a property under `data` does it work since `data` is now a POJO.

In the end the default behavior of Ractive is to operate data like POJOs. The warning is there because it may work in most cases, but you've been warned that it might break.

**Comment by evs-chris** [comment]
What @fskreuz said, plus there is `ractive.reset()` which probably won't work very well with adapted root data. So if you aren't using components or `reset`, you're probably good. Also note that adaptors are due for an overhaul in 0.10 to iron out some of the edgier cases, so swing by #2512 if you'd like to drop some input on usecases to support.

**Comment by wearhere** [comment]
Wow, thanks for the quick and comprehensive response @fskreuz and @evs-chris! I'll try to use POJOs for the root objects, don't want to fight the framework as I'm learning it. :) I'll take a look at #2512 too though it may take me a little to wrap my head around that. And speaking of learning the framework, as I mentioned on https://github.com/ractivejs/ractive/issues/2899, it's really great to see Ractive's community working so hard to figure out how it works and what could be made better.

**Comment by ghost** [comment]
But! If you're only after type-safety and a well-defined object structure, not everything else fancy about models (observers, accessors, hidden properties and all that fun stuff that warrants an adaptor), Ractive [_does_ accept a constructor or an instance of a constructor as `data`](https://ractive.js.org/API/Initialization%20Options/#data). It's treated like the function and plain object form of `data`, respectively.

[In this example](https://jsfiddle.net/m8yL9fkL/1/), it's close to the previous examples. But since the properties are public, Ractive can operate on them directly. And since there's no special sorcery to manipulate the values, adaptors aren't required.

You get the best of both worlds in terms of types as well as POJOs. You'll still get the warning though.
